### PR TITLE
internal/codegen: add Enum.Valid and AllEnumValues

### LINKED ISSUE #1607: autogenerate enum validation methods
### What do you want to change?

The codegen for enums generates a list of constants containing valid values:

```go
type EnumType string

const (
	EnumTypeA EnumType = "A"
	EnumTypeB EnumType = "B"
)
```

When writing developer tools, it can be helpful to convert from a string value to an enum type. However, there is no exhaustive list of valid enum values, so we have to wait until a query fails to discover that an invalid enum value has been provided.

I'd like the codegen to also emit something like:

```go
func (e EnumType) Valid() bool {
  switch e {
  case EnumTypeA, EnumTypeB:
    return true
  default:
    return false
  }
}
```

Then I can convert from string to enum type and check valid early on.

I know that this applies to PostgreSQL and Go. I'm not sure about MySQL and the other languages.

### What database engines need to be changed?

PostgreSQL

### What programming language backends need to be changed?

Go

### PR DESCRIPTION
**Pull request by josharian**
This commit adds a Valid method for enum types
and an All...Values function to get all enum values.

This makes it easier to work with enums.

Fixes #1607

### CODE REVIEW

**Review by kyleconroy — CHANGES_REQUESTED** [review]
Thanks for making a first pass at this. A few things before this is ready to merge. 

For enums with many values, the generated code takes up a large amount of horizontal space. Can we put each enum on its own line for both methods?

```go
func (e BooksBookType) Valid() bool {
	switch e {
	case BooksBookTypeFICTION:
		return true
	case BooksBookTypeNONFICTION:
		return true
	}
	return false
}

func AllBooksBookTypeValues() []BooksBookType {
	return []BooksBookType{
		BooksBookTypeFICTION,
		BooksBookTypeNONFICTION,
	}
}
```

This feature has only been requested by a few people. I'd like to put both behind a feature flag. Let's go with `emit_enum_valid_method` and `emit_all_enum_values`. I know the emit prefixes are a bit weird, but it's what we use for other options.

**Comment by josharian** [comment]
> Can we put each enum on its own line for both methods?

We can, although for the Valid method the compiler will generate better code if we can make them part of a single case statement, so it can unify the return statements. (That's also the reason not to write the code `return e == EnumA || e == EnumB || ... || e == EnumX`.)

What do you think about this:

```go
func (e BooksBookType) Valid() bool {
	switch e {
	case BooksBookTypeFICTION,
                 BooksBookTypeNONFICTION:
		return true
	}
	return false
}
```

**Comment by kyleconroy** [comment]
I think it's more straightforward to put one on each line. The comma style with a single case statement looks good to me.

**Comment by josharian** [comment]
OK, there's now one per line. Go's templates sure don't make that easy.

This is now gated by config parameters. There's a surprising amount required to add a config parameter; I hope I didn't miss anything.

There is also an additional commit to make debugging regeneration failures easier to diagnose. (I'd like to do https://twitter.com/commaok/status/1345977948453490688 as well, but this is a good first start.)

**Comment by jakoguta** [comment]
The test data `emit_enum_valid_and_values` added is for PostgreSQL. I recommend we put them in the folder `postgresql/pgx`.

**Comment by josharian** [comment]
> The test data emit_enum_valid_and_values added is for PostgreSQL. I recommend we put them in the folder postgresql/pgx.

Genuine question: Does it matter? This codegen change is DB-agnostic.

**Comment by josharian** [comment]
Ping.

What expectations should I have for review and issue comment turnaround time? It’d be useful to know so that I can plan accordingly and also not pester pointlessly.

File: scripts/regenerate/main.go [hunk 1]
```diff
@@ -21,9 +21,9 @@ func regenerate(dir string) error {
 			cwd := filepath.Dir(path)
 			cmd := exec.Command("sqlc-dev", "generate", "--experimental")
 			cmd.Dir = cwd
-			failed := cmd.Run()
+			out, failed := cmd.CombinedOutput()
 			if _, err := os.Stat(filepath.Join(cwd, "stderr.txt")); os.IsNotExist(err) && failed != nil {
-				return fmt.Errorf("%s: sqlc-dev generate failed", cwd)
+				return fmt.Errorf("%s: sqlc-dev generate failed\n%s", cwd, out)
```
Comment (kyleconroy) [inline scripts/regenerate/main.go:26]: out is a byte slice, so I think this will print out gibberish.

**Review by kyleconroy — APPROVED** [review]
One small issue in the regen script, going to merge anyways

File: scripts/regenerate/main.go (hunk 1 above)
Comment (josharian) [reply scripts/regenerate/main.go:26]: If the verb were `%v` then it would. With `%s` it prints as a string.

**Pull request merged**